great_expectations/datasource/fluent/sources.py

```diff
@@ -560,11 +562,11 @@ def add_or_update_datasource(
             id_: uuid.UUID | None = getattr(
                 self._data_context.datasources.get(datasource_name), "id", None
             )
-            if id_ and name_or_datasource:
-                if isinstance(name_or_datasource, str):
-                    kwargs["id"] = id_
-                else:
+            if id_:
+                if name_or_datasource and not isinstance(name_or_datasource, str):
                     name_or_datasource.id = id_
+                else:
+                    kwargs["id"] = id_
```

Comment:
This was the cause of the `name` positional args working but the keyword arg not working.